### ISSUE TITLE
sql: any DELETE upgrades a SNAPSHOT txn to SERIALIZABLE

### ISSUE THREAD

**Issue by andreimatei** [cockroachdb/cockroach#19876]
At the KV level, any `DelRange` request sets the `RetryOnPush` field on the txn, causing the transaction to return a retriable error if it is pushed (reason `RETRY_DELETE_RANGE`). For SNAPSHOT txns that means that they'll behave like SERIALIZABLE txns. Why? See [1].

SQL uses `DelRange` for deleting rows, even for deleting a single row with a known PK. So, as soon as a DELETE statement is used in a SNAPSHOT txn, the point of the SNAPSHOT txn goes out the window. Actually, it's even worse, as SQL does not perform eager retry for these pushed SNAPSHOT txns like it does for pushed SERIALIZABLE txns.

I believe SQL does not need to use `DelRange`. It could explicitly `Delete` every column family for the row(s) in question. It's not entirely clear to me if we'd need to `Delete` every col fam that could possibly exist or only the ones that actually exist (col fams with only NULL vals are not actually present in KV).
A concern that @danhhz brought up was about deletions performed during a schema change that removes a column family -> if we were to explicitly delete col fams, would the col fam in question be left as garbage and not cleaned up properly because the descriptor used for the ? I hope this is FUD and everything works fine due to the "delete_only" state that a col fam goes through during the schema change. But I haven't really thought about it.

[1] Details about why these txns need to restart can be seen in #6240. Essentially, since a `DelRange` only lays down intents for keys that exist, if the txn is pushed later on, it's possible for another txn to write a key that should be deleted but isn't, which is analogous to a lost update (I think), which is not permitted under snapshot isolation.

cc @vivekmenezes @tschottdorf @bdarnell 

Vivek, this came about because some people are complaining about seeing these `RETRY_DELETE_RANGE` restarts in SNAPSHOT txns, and I feel for them. The point of SNAPSHOT is quite severely put in question by this fact, and so I think we should do something about it.

**Comment by danhhz** [comment]
Ah, yup, I forgot about "delete_only". We should be fine then

**Comment by bdarnell** [comment]
When the `DELETE` specifies a single row by PK, we can definitely use `Delete` instead of `DeleteRange` and avoid this issue. But the more interesting question to me is what happens when the SQL `DELETE` specifies a range: `DELETE FROM t WHERE id >= $1 and id < $2`. If this changed from KV `DeleteRange` to a select-then-delete, it would introduce the same anomalies that #6240 tried to prevent. But these same anomalies are present in an `UPDATE` query with a similar `WHERE` clause. 

I think these anomalies are best understood as a kind of write skew (which is permitted by snapshot isolation) rather than as a kind of lost update (which is not). If we agree about this, we can just get rid of the `RetryOnPush` flag completely and let `DeleteRange` act the same as a scan followed by a delete.

**Comment by bdarnell** [comment]
> The point of SNAPSHOT is quite severely put in question by this fact, and so I think we should do something about it.

I also feel like exposing SNAPSHOT isolation was a premature optimization, so it's valid to question its usefulness. It is quite difficult to identify situations in which it would provide a performance benefit while still being safe.

**Comment by tschottdorf** [comment]
One of the founding principles in CockroachDB was that it's too difficult to reason about isolation levels. I just spent a few minutes with @andreimatei looking at the [critique of ANSI isolation levels](https://www.cs.umb.edu/cs734/CritiqueANSI_Iso.pdf) but unfortunately it doesn't quite cover what we need. A relevant quote is

> When the transaction T1 is ready to commit, it gets a
Commit-Timestamp, which is larger than any existing
Start-Timestamp or Commit-Timestamp. The transaction
successfully commits only if no other transaction T2 with a
Commit-Timestamp in T1’s interval [Start-Timestamp,
Commit-Timestamp] wrote data that T1 also wrote.

On a philosophical level, this raises the question of whether `DeleteRange` "writes" all the keys within its range. On a practical level, the answer is "yes" because SQL uses `DeleteRange`, and so through the SQL lens we can get lost update single table row updates (?). @andreimatei also points out that we could potentially violate SQL constraints that we care about (parent-children relationships), at which point CockroachDB as a system has a problem.

In the presence of this FUD, I'd hate to make `SNAPSHOT` less consistent.

For now, my take-away would be that SNAPSHOT isn't useful when deletions are part of the transaction. However, for a simple insert-only transactions, it should still be useful (caveats could be imposed by various index updates that at some level introduce range deletions during UPSERTs, etc).

Would be very interested in having a rigorous investigation of this.

**Comment by andreimatei** [comment]
Well, but where does this leave us wrt Ben's observation that the same anomaly is possible when doing a range `UPDATE`? Should we disallow any transaction that did any range operation from being pushed?

**Comment by tschottdorf** [comment]
Either we invest a considerable amount of time to identify the actual constraints of the (flavor of) SNAPSHOT isolation we can possibly support, or we get rid of snapshot (within sql, i.e. upgrade `SNAPSHOT` to `SERIALIZABLE` transparently). Priorities suggest the latter.

I also wonder how other databases actually fare in practice when it comes to constraints.

**Comment by bdarnell** [comment]
Write skew allows foreign key constraints to be violated: the constraints need to be checked at the transaction's write timestamp instead of its read timestamp (so they need to be re-validated at the new timestamp whenever the transaction gets pushed). Implementing this might also get us closer to deferred constraints. However, we don't want to invest much time in improving the performance of SNAPSHOT transactions, so for now I think we need to disallow the combination of SNAPSHOT and writes to tables affected by FK constraints (and multi-column CHECK constraints?), either by transparently upgrading the txn isolation, or maybe returning an error.

As for DeleteRange, if my analysis is correct and the same issue occurs with UPDATE, then I think upgrading txn isolation in `DeleteRange` is unnecessary. I think I'm OK with allowing this anomaly in SNAPSHOT mode. 

Should we get rid of SNAPSHOT altogether? I think probably not at this point, but I'm definitely more wary of it than before. Could there be more things like constraint checks that we're missing?
We should at least make sure we're not promoting it in the docs, and maybe map READ COMMITTED to SERIALIZABLE instead of SNAPSHOT (so you'd only get SNAPSHOT if you explicitly ask for it). Maybe we should go so far as disabling it completely.

**Comment by andreimatei** [comment]
> However, we don't want to invest much time in improving the performance of SNAPSHOT transactions, so for now I think we need to disallow the combination of SNAPSHOT and writes to tables affected by FK constraints (and multi-column CHECK constraints?), either by transparently upgrading the txn isolation, or maybe returning an error.

Umm I think if we are to do this, we might as well remove snapshot isolation and always use serializable... Wouldn't you say? cc @lego who I believe is doing some constraint checking tools.

> As for DeleteRange, if my analysis is correct and the same issue occurs with UPDATE, then I think upgrading txn isolation in DeleteRange is unnecessary. I think I'm OK with allowing this anomaly in SNAPSHOT mode.

Yeah... How about I go ahead and remove this `RetryOnPush` field?

**Comment by tschottdorf** [comment]
I don't think we should make any changes to how `SNAPSHOT` works at this point as it's not a priority and I don't think we should be allowing more anomalies in such an ad-hoc way.

**Comment by andreimatei** [comment]
But remember how this discussion started - as it stands, the SNAPSHOT isolation is useless as it gets upgraded too easily. And as Ben noted, "more anomalies" may count instances of anomalies, but not categories.

**Comment by tschottdorf** [comment]
all I'm saying is that I don't see a good reason to make what could be a pretty impactful change. Why don't you just upgrade SNAPSHOT to SERIALIZABLE instead? It's not clear to me which one is better though obviously I have a different favorite than you do. We don't understand snapshot isolation very well let alone its interplay with serializable, and so I would err on the side of caution and, in the absence of priority, inaction.

**Comment by andreimatei** [comment]
Well, a user is complaining that their expectations about SNAPSHOT (which we told them to use to avoid so many retries) were not met, and I happen to agree with them. So there's that for "priority".
And I guess I also question the "impactful change". I would imagine that, were I to do this change, there'd be no impact whatsoever and this will quickly subside from our minds. Modulo that constraint checking tool. But I guess we need to support inconsistencies in constrains anyway...

**Comment by tschottdorf** [comment]
And it seems that after this discussion, we may well want to advise that
user to stay away from SNAPSHOT. Either way, I don't have anything to add
here, so if @bdarnell is gung ho, then so be it.

> --

-- Tobias

**Comment by andreimatei** [comment]
I'm also interested in what @bdarnell has to say.
But I also want to make one more personal note: the simple thing to do here is, in my opinion, getting rid of `RetryOnPush` rather than doing nothing. That gives us a simpler codebase and product. And also a more... consistent product in the common use of the word - similar anomalies will consistently be possible, rather than the behavior differing based on which particular range operation one performs.

**Comment by bdarnell** [comment]
I think removing `RetryOnPush` at this point is a bad idea. If we decide to do implicit isolation upgrades in more places (one of the options from [this comment](https://github.com/cockroachdb/cockroach/issues/19876#issuecomment-342951971)), `RetryOnPush` may be the way we do it. Removing this field is a risk with no benefit. 

> as it stands, the SNAPSHOT isolation is useless as it gets upgraded too easily.

I disagree that this makes SNAPSHOT useless. Most transactions don't involve deletes. And even if we make FK checks (and potentially other things) also upgrade the transaction's isolation, there's still plenty of room for transactions to be written that can take advantage of SNAPSHOT mode. 

SNAPSHOT isolation may still be useless because it only helps (without introducing bugs) in fairly rare cases, but I don't believe that upgrading the priority on DELETE is a significant reducing in its usefulness. 

> Well, a user is complaining that their expectations about SNAPSHOT (which we told them to use to avoid so many retries) were not met, and I happen to agree with them. 

But we never promised that SNAPSHOT isolation avoids *all* retries, only some reduction. The real guarantee of SNAPSHOT mode is that we prevent every transactional anomaly except write skew. Everything else, including when and whether it may perform better than SERIALIZABLE is an implementation detail. Similarly, our READ COMMITTED actually gives you repeatable reads (and more), and we don't say that this fails to meet user's expectations about this isolation level. 

tl;dr? We should stop recommending SNAPSHOT mode even to users who want to minimize restarts. We may want to map READ COMMITTED to SERIALIZABLE instead of SNAPSHOT so that the only way to run SNAPSHOT is to explicitly ask for it. This would allow for experimentation so we could figure out whether it's worth having. But unless and until we're ready to rip out all of SNAPSHOT mode, we shouldn't touch things without a better reason than "simplification".

**Comment by andreimatei** [comment]
Mostly fair enough. But then, wouldn't it follow that we should set `RetryOnPush` on every (range)scan? IOW, is doing it just on `DelRange` defensible at all?

**Comment by bdarnell** [comment]
Setting RetryOnPush on every scan really would be tantamount to removing SNAPSHOT. If we believe that would be necessary, we should just remove SNAPSHOT instead.

I think that removing the assignment to RetryOnPush in evalDeleteRange might be a reasonable idea, except for the fact that it's a change to reduce consistency in a poorly-understood mode of operation and we should let the status quo remain unless we have a better reason to invest time in this area. Err on the side of consistency and inaction. 

Even if we removed this assignment, I would argue against removing the RetryOnPush feature because we've identified other cases in this thread where we might want to use implicit isolation upgrades and so we should probably leave it in to avoid migration headaches when/if we start using it for something else.

**Comment by bdarnell** [comment]
We're not planning to do anything about this. (If anything, we may remove SNAPSHOT entirely)
